### PR TITLE
V10/bugfix/14543 publish descendants

### PR DESCRIPTION
**Pull request by Migaroez**
### Reason
Fix for https://github.com/umbraco/Umbraco-CMS/issues/14543

### Description
- Renamed vars to clear up some things
- Skipped the branchitem change detection on the root node as the action implies we want to save and publish the root
- Seperated the root and children publish results to have seperate references to them later on
- Used the seperated root reference to properly check if the root was succesful after all the children are processed.
- Passed the state of root from the publishing to the published notification properly linking the pair together

### Reproduction steps
Add the following classes to your project
```csharp
using Umbraco.Cms.Core.Composing;
using Umbraco.Cms.Core.Events;
using Umbraco.Cms.Core.Notifications;

public class Published : INotificationAsyncHandler<ContentPublishedNotification>
{
    public Task HandleAsync(ContentPublishedNotification notification, CancellationToken cancellationToken)
    {
        notification.State.TryGetValue("publishing", out var value);

        return Task.CompletedTask;
    }
}
public class Publishing : INotificationAsyncHandler<ContentPublishingNotification>
{
    public Task HandleAsync(ContentPublishingNotification notification, CancellationToken cancellationToken)
    {
        notification.State.Add("publishing", notification.PublishedEntities.Select(x => x.Id).ToList());

        return Task.CompletedTask;
    }
}

public class Saved : INotificationAsyncHandler<ContentSavedNotification>
{
    public Task HandleAsync(ContentSavedNotification notification, CancellationToken cancellationToken)
    {
        notification.State.TryGetValue("saving", out var value);

        return Task.CompletedTask;
    }
}
public class Saving : INotificationAsyncHandler<ContentSavingNotification>
{
    public Task HandleAsync(ContentSavingNotification notification, CancellationToken cancellationToken)
    {
        notification.State.Add("saving", notification.SavedEntities.Select(x => x.Id).ToList());

        return Task.CompletedTask;
    }
}

public class MyComposer : IComposer
{
    public void Compose(IUmbracoBuilder builder)
    {
        builder.AddNotificationAsyncHandler<ContentPublishingNotification, Publishing>();
        builder.AddNotificationAsyncHandler<ContentPublishedNotification, Published>();
        builder.AddNotificationAsyncHandler<ContentSavingNotification, Saving>();
        builder.AddNotificationAsyncHandler<ContentSavedNotification, Saved>();
    }
}
```

Each pair (save or publish) should be hit for every child that has unpublished changes when executing publish with descendants on a parent (or ancestor for that matter).
These pairs should also be called on the parent
And the state should be passed between each item in a pair.

### CODE REVIEW

File: src/Umbraco.Core/Services/ContentService.cs [hunk 1]
```diff
@@ -2164,7 +2168,12 @@ public IEnumerable<PublishResult> SaveAndPublishBranch(IContent content, bool fo
             // (SaveAndPublishBranchOne does *not* do it)
             scope.Notifications.Publish(
                 new ContentTreeChangeNotification(document, TreeChangeTypes.RefreshBranch, eventMessages));
-            scope.Notifications.Publish(new ContentPublishedNotification(publishedDocuments, eventMessages));
+            if (rootResult?.Success == true)
```
Comment (Zeegaan) [inline src/Umbraco.Core/Services/ContentService.cs:2171]: Pattern matching is the new hotness 🔥 
```suggestion
            if (rootResult?.Success is true)
```

File: src/Umbraco.Core/Notifications/NotificationExtensions.cs [hunk 2]
```diff
@@ -1,3 +1,5 @@
+using Umbraco.Cms.Core.Models;
+
 namespace Umbraco.Cms.Core.Notifications;
```
Comment (Zeegaan) [inline src/Umbraco.Core/Notifications/NotificationExtensions.cs:3]: This seems unnecessary 😁 
```suggestion
namespace Umbraco.Cms.Core.Notifications;
```

**Review by Zeegaan — CHANGES_REQUESTED** [review]
Found a few small things 👀

File: src/Umbraco.Core/Notifications/NotificationExtensions.cs (hunk 2 above)
Comment (Migaroez) [reply src/Umbraco.Core/Notifications/NotificationExtensions.cs:3]: 🙈 leftover i forgot to clean up

**Comment by Zeegaan** [comment]
Looks good, tests good!

**Pull request merged**